platform/socket.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "std/string.hpp"
+#include "std/target_os.hpp"
+
+#if defined(OMIM_OS_IPHONE) || defined(OMIM_OS_MAC)
+@class SocketImpl;
+#else
+class SocketImpl;
+#endif
+
+namespace platform
+{
+class Socket
+{
+public:
+  Socket();
+  ~Socket();
+
+  bool Open(string const & host, uint16_t port);
+  void Close();
+
+  bool Read(uint8_t * data, uint32_t count);
+  bool Write(uint8_t const * data, uint32_t count);
+
+  void SetTimeout(uint32_t milliseconds);
+
+private:
+  SocketImpl * m_socketImpl = nullptr;
```

Comment:
In general, I dislike the idea of PIMPL because it's hard to mock sockets, and, as consequence, it's hard to test code that uses sockets. Why not to create a base abstract class, with a single real implementation and a lot of mock implementations. All users of sockets API should use a SocketsFactory, and tests should be able to replace this default SocketFactory.